File: frontend/src/api/discordPresence.ts
import { apiFetchJson, apiFetchVoid } from './client'
import type { DiscordPresenceStatus } from '../types'

export type DiscordPresenceActivityPayload = {
  video_id: string
  title: string
  channel_title?: string | null
  playlist_name?: string | null
  source_url?: string | null
  is_playing: boolean
  is_playlist_playback: boolean
  position_seconds: number
  duration_seconds?: number | null
}

export async function fetchDiscordPresenceStatus() {
  return apiFetchJson<DiscordPresenceStatus>('/api/discord-presence', {
    method: 'GET',
  })
}

export async function updateDiscordPresenceActivity(
  payload: DiscordPresenceActivityPayload,
) {
  return apiFetchJson<DiscordPresenceStatus>('/api/discord-presence/activity', {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

export async function clearDiscordPresenceActivity() {
  return apiFetchVoid('/api/discord-presence/activity', {
    method: 'DELETE',
  })
}

File: backend/app/models/discord_presence.py
from pydantic import BaseModel


class DiscordPresenceStatus(BaseModel):
    enabled: bool
    configured: bool
    available: bool
    connected: bool
    active: bool
    last_error: str | None = None


class DiscordPresenceActivityRequest(BaseModel):
    video_id: str
    title: str
    channel_title: str | None = None
    playlist_name: str | None = None
    source_url: str | None = None
    is_playing: bool = True
    is_playlist_playback: bool = False
    position_seconds: int = 0
    duration_seconds: int | None = None
